diff --git a/ux_elements.module b/ux_elements.module
index 1b86d8e..f4ce84a 100644
--- a/ux_elements.module
+++ b/ux_elements.module
@@ -13,7 +13,7 @@ function ux_elements_element_info() {
     '#theme_wrappers' => array('horizontal_tabs'),
     '#default_tab' => '',
     '#pre_render' => array('form_pre_render_fieldset'),
-    '#process' => array('form_process_horizontal_tabs'),
+    '#process' => array('form_process_horizontal_tabs2'),
   );
   $types['vertical_accordion'] = array(
     '#theme_wrappers' => array('vertical_accordion', 'fieldset'),
@@ -29,7 +29,7 @@ function ux_elements_element_info() {
  */
 function ux_elements_theme() {
   return array(
-    'horizontal_tabs' => array(
+    'horizontal_tabs2' => array(
       'render element' => 'element',
     ),
     'vertical_accordion' => array(
@@ -52,7 +52,7 @@ function ux_elements_theme() {
  * @return
  *   The processed element.
  */
-function form_process_horizontal_tabs($element, &$form_state) {
+function form_process_horizontal_tabs2($element, &$form_state) {
   // Inject a new fieldset as child, so that form_process_fieldset() processes
   // this fieldset like any other fieldset.
   $element['group'] = array(
@@ -73,7 +73,7 @@ function form_process_horizontal_tabs($element, &$form_state) {
  *
  * @param $variables
  */
-function theme_horizontal_tabs($variables) {
+function theme_horizontal_tabs2($variables) {
   $element = $variables['element'];
 
   $output = '<h2 class="element-invisible">' . t('Horizontal Tabs') . '</h2>';